### PR TITLE
campaign testing functional

### LINKED ISSUE #24: Create Tests for Campaign endpoints
Using pytest and unit test 
test the endpoints for Campaigns
`create` --> `POST /api/v1/campaigns`
`update`--> `PUT /api/v1/campaigns/<int:campaign_id>`
`delete`--> `DELETE /api/v1/campaigns/<int:campaign_id>`
`index`--> `GET /api/v1/campaigns`
`show`--> `GET /api/v1/campaigns/<int:campaign_id>`

### PR DESCRIPTION
**Pull request by perryr16**
## What's this PR do?
* Create tests for campaign API endpoints

## Where should the reviewer start?
* `eth_crowdfund_api/src/tests/test_campaign.py`

## How should this be manually tested?
* run tests from main folder OR inside the pipenv shell 
* create the test database `createdb eth_crowdfund_api_db_test`
* install pytest and coverage `pip install pytest pytest-cov`
* run command `pytest`
* for coverage run `pytest --cov=src`
  * if that does not work, `pip install pytest-cov`
* if there are errors about modules not being found, make sure they are installed inside AND outside of the pipenv shell

## Any background context you want to provide?
* testing somewhat followed form this article
* https://www.codementor.io/@olawalealadeusi896/building-a-restful-blog-apis-using-python-and-flask-part-3-lx7rt8pfk

## What are the relevant tickets?
close #24

### CODE REVIEW

File: eth_crowdfund_be/eth_crowdfund_api/src/config.py [hunk 1]
```diff
@@ -14,8 +14,15 @@ class Production(object):
     SQLALCHEMY_DATABASE_URI = "postgresql://postgres:password@localhost/eth_crowdfund_api_db"
     JWT_SECRET_KEY = os.getenv('JWT_SECRET_KEY')
 
+class Testing(object):
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/config.py:17]: create a testing enviornment

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_sample.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+def func(x):
+  return x + 1 
+
+def test_answer():
+  assert func(3) == 4
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_sample.py:5]: This is a basic test that should pass without any database or app involvement. to run individually its `pytest <path to file depending on current directory>`  ex: `pytest src/tests/test_sample.py`

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+def func(x):
+  return x + 1
+# basic test
+def test_answer():
+  assert func(3) == 4
+
+class CampaignTest(unittest.TestCase):
+  # set up the tests with 2 campaigns
+  # This runs before EVERY test. not once before all
+  def setUp(self):
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py:18]: There are some comments through out that should help with readability. This setup will run before each test

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        "manager": "13",
+        "address": "14",
+        "min_contribution": 50.2
+    }
+    
+    with self.app.app_context():
+      # create all db objects
+      db.create_all()
+      # delete test campaigns by name
+      existing_campaign1 = Campaign.get_campaign_by_name(self.campaign1["name"])
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py:46]: delete campaigns before each test so they can be created again

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  def test_campaign_update(self):
+    campaign1_update = {
+      "description": "new description", 
+      "contributors": 5,
+      "manager": "1x3y"
+    }
+    # create the user before updating
+    res = self.client().post('/api/v1/campaigns/',
+                             headers={'Content-Type': 'application/json'},
+                             data=json.dumps(self.campaign1))
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py:81]: This test could be more DRY, but its a first pass

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+
+    res_index = self.client().get('/api/v1/campaigns/',
+                   headers={'Content-Type': 'application/json'})
+    json_data = json.loads(res_index.data)
+    self.assertEqual(len(json_data), 2)
+    self.assertEqual(res_index.status_code, 200)
+
+  # def test_get_one_campaign(self):
+  #   # create 2 campaigns
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py]: Not sure why this test wouldnt pass. When sending a request to `get('/api/v1/campaigns/<campaign_id>...` I was getting a status code of 308 which is a `permanent redirect message`

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    self.assertNotEqual(self.campaign1["min_contribution"], 100)
+    self.assertEqual(res.status_code, 201)
+
+  def test_campaign_update(self):
+    campaign1_update = {
+      "description": "new description", 
+      "contributors": 5,
+      "manager": "1x3y"
+    }
+    # create the user before updating
```
Comment (AntonioJacksonII) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py]: Is this leftover from a copy & paste? Is it referring to creating a campaign before updating?

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    }
+    
+    with self.app.app_context():
+      # create all db objects
+      db.create_all()
+      # delete test campaigns by name
+      existing_campaign1 = Campaign.get_campaign_by_name(self.campaign1["name"])
+      if existing_campaign1:
+        existing_campaign1.delete()
+      existing_campaign2 = Campaign.get_campaign_by_name(self.campaign2["name"])
```
Comment (AntonioJacksonII) [inline eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py:49]: Did you create `get_campaign_by_name`? Where is it located?

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py (hunk 7 above)
Comment (perryr16) [reply eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py]: not copy-paste so much as looking at a word and trying to think of another. I was looking at a tutorial for creating users. I'll modify it to say campaigns

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py (hunk 6 above)
Comment (AntonioJacksonII) [reply eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py]: Did you try `get('/api/v1/campaigns/<campaign_id>/`? Sometimes I have success adding the final `/` when testing.

File: eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py (hunk 8 above)
Comment (perryr16) [reply eth_crowdfund_be/eth_crowdfund_api/src/tests/test_campaign.py:49]: `get_campaign_by_name` is a static method thats created in the Campaign model (`src/models/campaign.py`)

File: eth_crowdfund_be/eth_crowdfund_api/src/config.py [hunk 9]
```diff
@@ -14,8 +14,15 @@ class Production(object):
...
+  TESTING = True
+  JWT_SECRET_KEY = os.getenv('JWT_SECRET_KEY')
+  SQLALCHEMY_DATABASE_URI = "postgresql://postgres:password@localhost/eth_crowdfund_api_db"
```
Comment (takeller) [inline eth_crowdfund_be/eth_crowdfund_api/src/config.py]: We should probably set up a testing database 'eth_crowdfund_api_test_db'

Comment (perryr16) [reply eth_crowdfund_be/eth_crowdfund_api/src/config.py]: good thought. I'll change this

File: eth_crowdfund_be/eth_crowdfund_api/src/views/CampaignView.py [hunk 10]
```diff
@@ -29,7 +29,7 @@ def custom_response(res, status_code):
     status=status_code
   )
 
-@campaign_api.route('/<int:campaign_id>/', methods=['GET'])
+@campaign_api.route('/<int:campaign_id>', methods=['GET'])
```
Comment (perryr16) [inline eth_crowdfund_be/eth_crowdfund_api/src/views/CampaignView.py:32]: Antonio with the save

**Review by AntonioJacksonII — APPROVED** [review]
Great work getting testing set up! Testing is IMPORTANT.

**Review by takeller — APPROVED** [review]

**Pull request merged**